source/common/config/config_provider_impl.h

```diff
@@ -0,0 +1,370 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/config/config_provider.h"
+#include "envoy/config/config_provider_manager.h"
+#include "envoy/init/init.h"
+#include "envoy/server/admin.h"
+#include "envoy/server/config_tracker.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/thread_local/thread_local.h"
+
+#include "common/config/utility.h"
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+// This file provides a set of base classes, (StaticConfigProviderImplBase,
+// DynamicConfigProviderImplBase, ConfigProviderManagerImplBase, ConfigSubscriptionInstanceBase),
+// conforming to the ConfigProvider/ConfigProviderManager interfaces, which in tandem provide a
+// framework for implementing static and dynamic configuration for Envoy.
+//
+// Dynamic configuration is distributed via xDS APIs (see
+// https://github.com/envoyproxy/data-plane-api/blob/master/XDS_PROTOCOL.md). The framework exposed
+// by these classes simplifies creation of client xDS implementations following a shared ownership
+// model, where according to the config source specification, a config subscription, config protos
+// received over the subscription and the subsequent config "implementation" (i.e., data structures
+// and associated business logic) are shared across ConfigProvider objects and Envoy worker threads.
+//
+// This approach enables linear memory scalability based primarily on the size of the configuration
+// set.
+//
+// A blueprint to follow for implementing static and dynamic config providers is as follows:
+//
+// For both:
+//   1) Create a class derived from ConfigProviderManagerImplBase and implement the required
+//   interface.
+//      When implementing createXdsConfigProvider(), it is expected that getSubscription<T>() will
+//      be called to fetch either an existing ConfigSubscriptionInstanceBase if the config source
+//      configuration matches, or a newly instantiated subscription otherwise.
+//
```

Comment:
Yeah, there is definitely value in doing this although I've been primarily focused on the DynamicConfigProvider use case for this PR since that's where I suspect most of the benefit will be seen.

I'll add a TODO here and come back to it at the end of this PR chain.

Reply:
There were a couple thing to note from SDS implementation:
- resources in static_resources can be already referred by name, I know this is not the case for RDS.
- if we want to deduplicating inline config, be careful about DataSource, Envoy doesn't watch if the file content changed, so same proto may result different configuration at the end.

Reply:
> There were a couple thing to note from SDS implementation:
> 
> * resources in static_resources can be already referred by name, I know this is not the case for RDS.

Right, this is something we'll need to keep track of for the implementation of static resource sharing. Thanks for calling this out.

> * if we want to deduplicating inline config, be careful about DataSource, Envoy doesn't watch if the file content changed, so same proto may result different configuration at the end.

Noted. I'll file an issue for static/inline deduplication/sharing and will add a note for both of these items.